SwiftUI support

fredyshox opened the pull request:
Fixes #42 

Resolved by defining `PreferencePaneHostingController`, wrapper around `NSHostingController` with `PreferencePane` support. User provides custom SwiftUI view and data required by the protocol. Also I've added example preference UI `UserAccountsView`. 

I think this is optimal way of doing this, because we can just stick with existing API. Let me know if you have any thoughts on this.


<!-- Issuehunt content -->

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#42: SwiftUI support](https://issuehunt.io/repos/139150459/issues/42)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>
<!-- /Issuehunt content-->

sindresorhus commented:
Awesome! Thanks for working on this 🙌

I was hoping to keep as much as possible in SwiftUI, including providing the required data.

One idea could be something like this:

```swift
protocol PreferencePaneView: View {
	var preferencePaneIdentifier: Identifier { get }
	var preferencePaneTitle: String { get }
	var toolbarItemIcon: Image { get }
}

struct UserAccountsView: PreferencePaneView {
	var preferencePaneIdentifier = .userAccounts
	var preferencePanelTitle = "User Acccounts"
	var toolbarItemIcon = Image(…)

	var body: some View {}
}
```

sindresorhus commented:
Would also be nice to be able to use `Identifiable` somehow.

sindresorhus commented:
Actually, maybe something like this would be nicer:

```swift
struct UserAccountsView: PreferencePaneView {
	var preferencePaneConfig = .init(
		identifier: .userAccounts,
		title: "User Acccounts",
		icon: Image(…)
	)

	var body: some View {}
}
```

sindresorhus commented on Example/UserAccountsView.swift:
```diff
@@ -0,0 +1,134 @@
+                Picker(selection: self.$selection, label: EmptyView()) {
+                     Text("Automatic").tag(0)
+                     Text("Manual").tag(1)
+                }.fixedSize()
```
This dropdown changes size when you choose each item. Would be better to hard-code the size or even better find the longest label and use that.

sindresorhus commented on Example/UserAccountsView.swift:
```diff
@@ -0,0 +1,134 @@
+                VStack() {
+                    Toggle("Allow user to administer this computer", isOn: self.$isOn1)
+                        .toggleStyle(CheckboxToggleStyle())
+                    Text("Administrator has root access to this machine")
```
```suggestion
                    Text("Administrator has root access to this machine.")
```

fredyshox commented on Example/UserAccountsView.swift:
```diff
@@ -0,0 +1,134 @@
+            RadioSectionView(selection: self.$selection1)
+            Divider()
+                .frame(height: 20.0)
+                .alignmentGuide(.preferenceLabel, computeValue: { $0[.leading] + 104.0 })
```
Its magic number, necessary to spread Divider across width of the container 😄. It's actually the width of the widest title (label on the left of each section). I can try to create custom Text with Binding to get its width, and then use `max(widths)` to calculate this offset.

fredyshox commented:
@sindresorhus 

Hey I was thinking about what you said about bundling some components for simple panes, and came up with and idea to create custom `Preferences` and `PreferenceSection` components. They do all the alignment and layout magic, so user doesn’t need to know anything about custom AlignmentGuides etc.. 

I’ve also decided to define `PreferencePaneView`, just as you proposed.

sindresorhus commented on Sources/Preferences/SwiftUIComponents.swift:
```diff
@@ -0,0 +1,128 @@
+*/
+@available (macOS 10.15, *)
+@_functionBuilder
+public struct SectionBuilder {
```
This name is too generic for being public.
```suggestion
public struct PreferenceSectionBuilder {
```

sindresorhus commented on Example/UserAccountsView.swift:
```diff
@@ -44,86 +27,45 @@ extension Font {
-					.font(.preferenceDescription)
-					.foregroundColor(.secondary)
-				Toggle("Allow user to access every file", isOn: $isOn2)
+			PreferenceSection(title: "Show scroll bars:") {
```
Like with most other views, `PreferencesSection` should accept both a title and a view. This is necessary when you need something else than just `Text()` for the title. Example:

<img width="780" alt="Screenshot 2020-02-11 at 23 12 53" src="https://user-images.githubusercontent.com/170270/74255148-0a963900-4d24-11ea-87af-5db64d3374e4.png">

sindresorhus commented on Example/UserAccountsView.swift:
```diff
@@ -44,86 +27,45 @@ extension Font {
-				Text("When scrolling").tag(0)
-				Text("Always").tag(1)
+		Preferences(contentWidth: contentWidth) {
+			PreferenceSection(title: "Permissions:") {
```
To reduce globals, I think it would be nice to nest these things under the `Preferences` namespace using typealiases.

- `Preferences.Container`
- `Preferences.Section`

sindresorhus replied:
We could maybe also do this for `PreferencesPaneView` => `Preferences.PaneView`.

sindresorhus commented on Sources/Preferences/SwiftUIComponents.swift:
```diff
@@ -0,0 +1,128 @@
+    public init<Content: View>(title: String, bottomDivider: Bool = true, content: @escaping () -> Content) {
+        self.title = title
+        self.bottomDivider = bottomDivider
+        self.content = { AnyView(content()) }
```
Do we have to use `AnyView` here?

fredyshox replied:
In order for PreferenceSectionBuilder to work and output array of sections, I had to type-erase the content.

sindresorhus commented on Sources/Preferences/SwiftUIComponents.swift:
```diff
@@ -0,0 +1,128 @@
+        }
+    }
+	
+	var fontSize: CGFloat {
```
```suggestion
	let fontSize = 13.0
```

sindresorhus commented:
Hey, really sorry about the lack of response. This simply got lost in all my tabs.

It looks very good now. 👌

One thing I've realized is that we cannot currently ship SwiftUI stuff directly because of [this dumb bug](https://github.com/feedback-assistant/reports/issues/44). Could you make a separate SwiftUI target for this for Swift Package Manager?

Only allowing `PreferencesSection` inside `Preferences` will probably be a bit too limiting. What if you need some custom view between two of the sections?

And just so we don't forget, this will also need some docs when we've finalized how it should work.

fredyshox commented on Sources/Preferences/SwiftUIComponents.swift:
```diff
@@ -0,0 +1,128 @@
+@available (macOS 10.15, *)
+public struct Preferences: View {
+    public let sectionBuilder: () -> [PreferenceSection]
+	public let contentWidth: CGFloat
```
I switched to Double in all public APIs, leaving CGFloat's in internal implementations.

fredyshox commented:
Ok so it seems to be complete. Here's description of major recent changes:
* I've found solution using custom PreferenceKey and overlay with GeometryReader to find max width of label components, so I could remove workaround using NSTextField. 
* Following your suggestion I've added convenience initializer for custom View instead of Text as label component
* Added separate PreferencesSwiftUI target as library in SwiftPM 
* Fixed issues with weak linking in CocoaPods and Carthage by adding `-weak_framework` flags (and tested on Mojave)

And as for the AnyView's, I don't see any reasonable alternative with ability to store Views in array. We could create custom ViewBuilder using TupleView, but it would require 10+ overloads, which is a bit ugly.

Also I've added section in README documenting usage. Let me know, if there's anything else 🙂

sindresorhus commented:
Alright. I'm finally done with travelling and other things, so I can focus on this PR again.

sindresorhus commented:
I think this was missed: https://github.com/sindresorhus/Preferences/pull/46/files#r377738897

sindresorhus commented:
While `PreferencePaneHostingController` is nice when you want to mix normal panes and SwiftUI panes or if you need more control, I think we should also have a simpler way if you only want SwiftUI panes. I'm thinking an overload init for `PreferencesWindowController` that accepts an array of SwiftUI views instead of `PreferencePane` for the `preferencePanes` parameter.

sindresorhus commented on Sources/PreferencesSwiftUI/PreferenceSection.swift:
```diff
@@ -0,0 +1,97 @@
+		self.content = content().eraseToAnyView()
+	}
+
+	public init<Content: View>(title: String, bottomDivider: Bool = true, content: @escaping () -> Content) {
```
Needs doc comments.

sindresorhus commented:
I have been thinking about this, and I think we should go with one config property like suggested earlier:

```swift
struct UserAccountsView: PreferencePaneView {
	var preferencePaneConfig = .init(
		identifier: .userAccounts,
		title: "User Accounts",
		icon: Image(…)
	)

	var body: some View {}
}
```

That way the required config doesn't get so mixed up with the view-specific properties.

sindresorhus commented on Sources/PreferencesSwiftUI/PreferencePane+SwiftUI.swift:
```diff
@@ -0,0 +1,42 @@
+public protocol PreferencePaneView: View {
+	var preferencePaneIdentifier: PreferencePane.Identifier { get }
+	var preferencePaneTitle: String { get }
+	var toolbarItemIcon: NSImage { get }
```
```suggestion
	var toolbarItemIcon: Image { get }
```

fredyshox replied:
But there's a problem, I don't see possibility to convert Image back to NSImage. And NSImage is required in `ToolbarItemStyleViewController` and for coexistance with AppKit's NSViewControllers

sindresorhus replied:
That's true. Oh, never mind.

sindresorhus commented on Sources/PreferencesSwiftUI/PreferencePane+SwiftUI.swift:
```diff
@@ -0,0 +1,42 @@
+/**
+	SwiftUI equivelent of PreferencePane protocol
+*/
+@available (macOS 10.15, *)
```
```suggestion
@available(macOS 10.15, *)
```

sindresorhus commented on readme.md:
```diff
@@ -180,6 +180,42 @@ As with any `NSWindowController`, call `NSWindowController#close()` to close the
+	PreferencePaneHostingController(preferencePaneView: CustomView())
+}
+
+struct CustomPane: View, PreferencePaneView {
```
```suggestion
struct CustomPane: PreferencePaneView {
```

sindresorhus commented on Example/UserAccountsView.swift:
```diff
@@ -0,0 +1,91 @@
+/**
+	Main view of User Accounts preference screen
+ */
+struct UserAccountsView: View, PreferencePaneView {
```
```suggestion
struct UserAccountsPane: PreferencePaneView {
```

?

sindresorhus commented on Example/UserAccountsView.swift:
```diff
@@ -0,0 +1,91 @@
+}
+
+/**
+	Main view of User Accounts preference screen
```
```suggestion
	The main view of “User Accounts” preference pane.
```

sindresorhus commented on Sources/PreferencesSwiftUI/PreferenceContainer.swift:
```diff
@@ -0,0 +1,77 @@
+	public let contentWidth: Double
+	@State private var maxLabelWidth: CGFloat = 0.0
+
+	public init(contentWidth: Double, @PreferenceSectionBuilder builder: @escaping () -> [PreferenceSection]) {
```
Needs a doc comment that documents the `contentWidth` parameter.

sindresorhus commented on Sources/PreferencesSwiftUI/PreferenceContainer.swift:
```diff
@@ -0,0 +1,77 @@
+}
+
+/**
+	Container for PreferenceSection objects.
```
This should say what it is and does, not just what it accepts.

sindresorhus commented on Example/UserAccountsView.swift:
```diff
@@ -0,0 +1,91 @@
+			}
+			PreferenceSection(title: "Preview mode:") {
+				VStack(alignment: .leading) {
+					Picker(selection: self.$selection2, label: EmptyView()) {
```
This should have a separate `$selection` binding.

sindresorhus commented on readme.md:
```diff
@@ -180,6 +180,42 @@ As with any `NSWindowController`, call `NSWindowController#close()` to close the
+
+If your deployment target is macOS 10.15 or later, you can use the bundled SwiftUI components to create panes. Your preference pane views must conform to `PreferencePaneView`, which is equivalent to `PreferencePane`. Then wrap the views in `PreferencePaneHostingController` and pass them to `PreferencesWindowController` as you would with normal panes.
+
+It also comes with `PreferenceContainer` and `PreferenceSection` to automatically achieve similar alignment as AppKit's [`NSGridView`](https://developer.apple.com/documentation/appkit/nsgridview).
```
Link `PreferenceContainer` and `PreferenceSection` to their source.

sindresorhus commented:
I think the doc comments for the main exported parts `PreferencePaneView`, `PreferenceContainer`, and `PreferenceSection` could be expanded and improved. Users will deal with these a lot, so having clear docs for them is important. Would also be nice to have an example in their doc comments too.

sindresorhus commented:
NSImage has:

```swift
/* These images are intended for use in toolbars in preference windows.
     */
    @available(OSX 10.5, *)
    public class let userAccountsName: String

    @available(OSX 10.5, *)
    public class let preferencesGeneralName: String

    @available(OSX 10.5, *)
    public class let advancedName: String
```

I think we should wrap them in `Image()` (SwiftUI) and expose them as extensions on `Image`, so the user can easily use them.

sindresorhus commented:
@fredyshox Friendly bump :)

fredyshox commented:
> I think this was missed: https://github.com/sindresorhus/Preferences/pull/46/files#r377738897

Wrap everything into namespace using `enum`?

sindresorhus commented:
Yup

fredyshox commented on Example/UserAccountsView.swift:
```diff
@@ -0,0 +1,91 @@
+			PreferenceSection(label: {
+				Toggle("Some toggle", isOn: self.$isOn3)
+			}) {
+				VStack(alignment: .leading) {
```
Hmm, I think its gonna be too limiting for users with custom layouts. And still it's just a one line for those who want default look.

sindresorhus replied:
I think we should optimize for the common-case. People that want custom layout can just use `VStack {}` themselves to get a centered layout.

fredyshox commented:
I’ve seen this #51 , and this make me think, that in SwiftUI I often apply modifiers outside of View definition, for example: `environmentObject` in most cases will be applied with some state-managing object, upon creation of `PreferencesWindowController` in AppDelegate. So requiring `Preferences.PaneView` conformance can be limiting. 

```swift
// …somewhere in AppDelegate
let view = 
    UserAccountsView()
        .environmentObject(self.accountManager)
// view type is now ModifierContent<UserAccountsView, Modifier>

let pane = Preferences.PaneHostingController(preferencePaneView: view) // error: view is not Preferences.PaneView
```

We can do this differently, by defining `Preferences.PaneConfig` to satisfy toolbar requirements and using it like this:

```swift
// …somewhere in AppDelegate
let config = Preferences.PaneConfig(
    identifier: "...",
    title: "User Accounts"
    toolbarIcon: NSImage(names: "img-name"))

let view = 
    UserAccountsView() // just plain view, not Preferences.PaneView
        .environmentObject(self.accountManager)
        .environmentObject(self.someOtherObject)
        .otherModifier()

let pane = Preferences.PaneHostingController(config: config, paneView: view)
```

@sindresorhus What do you think about this?

sindresorhus commented:
Good point.

Could we make it a view modifier? The benefit would be that the user would be able to  set/change the title and toolbar icon directly from SwiftUI.

```swift
let view = 
    UserAccountsView() // just plain view, not Preferences.PaneView
        .environmentObject(self.accountManager)
        .environmentObject(self.someOtherObject)
        .otherModifier()
        .preferencePane(
             identifier: "...",
             title: "User Accounts",
             toolbarIcon: NSImage(names: "img-name"))
        )

let pane = Preferences.PaneHostingController(paneView: view)
```

Another benefit is that it would work with https://github.com/sindresorhus/Preferences/pull/46#issuecomment-609729282.

sindresorhus commented:
Another solution would be:

```swift
let view = 
	Preferences.PaneView(
		identifier: "...",
		title: "User Accounts",
		toolbarIcon: NSImage(named: "img-name"))
	) {
		UserAccountsView()
			.environmentObject(self.accountManager)
			.environmentObject(self.someOtherObject)
			.otherModifier()
	}

let pane = Preferences.PaneHostingController(paneView: view)
```

The benefit of this is that it's more type-safe as we're forcing the user to pass the config and we only accept `Preferences.PaneView`-type views.

fredyshox commented:
Modifier would also be type-safe, only catch is it'd need to be the last applied `ViewModifier` in the chain:
```swift
// Preferences.PaneHostingController
init<T: View>(paneView: ModifiedContent<T, PaneConfigModifier>) {
    // ...
    paneView.modifier //... identifier, title, toolbarIcon
}
```

I guess we can do both 😃

sindresorhus commented:
I don't think we should do both. There should be one consistent way to do it.

sindresorhus commented:
> Modifier would also be type-safe, only catch is it'd need to be the last applied ViewModifier in the chain:

Yes, but then we need to document the fact that it needs to be last, and the user might accidentally add more modifiers and get a cryptic compile error.

sindresorhus commented:
I personally prefer `Preferences.PaneView()`. Which do you prefer (and why)?

fredyshox commented:
> > Modifier would also be type-safe, only catch is it'd need to be the last applied ViewModifier in the chain:
> 
> Yes, but then we need to document the fact that it needs to be last, and the user might accidentally add more modifiers and get a cryptic compile error.

I see, that can be a problem. 

So `Preferences.PaneView()` is fine for me.

sindresorhus commented:
> So Preferences.PaneView() is fine for me.

👍 Let's go with that then.

fredyshox commented:
@sindresorhus Done 👍

sindresorhus commented:
I think the only thing missing now is:

> While PreferencePaneHostingController is nice when you want to mix normal panes and SwiftUI panes or if you need more control, I think we should also have a simpler way if you only want SwiftUI panes. I'm thinking an overload init for PreferencesWindowController that accepts an array of SwiftUI views instead of PreferencePane for the preferencePanes parameter.

sindresorhus commented on Example/UserAccountsView.swift:
```diff
@@ -0,0 +1,92 @@
+import Preferences
+
+/**
+Function wrapping SwiftUI into PreferencePane, which is mimicing view controller's default construction syntax
```
```suggestion
Function wrapping SwiftUI into `PreferencePane`, which is mimicking view controller's default construction syntax.
```

sindresorhus commented on Example/UserAccountsView.swift:
```diff
@@ -0,0 +1,92 @@
+	/// Wrap your custom view into `Preferences.PaneView`, while providing necessary toolbar info.
+	let paneView =
+		Preferences.PaneView(
+			identifier: PreferencePaneIdentifier.userAccounts,
```
```suggestion
			identifier: .userAccounts,
```

sindresorhus commented on Example/UserAccountsView.swift:
```diff
@@ -0,0 +1,92 @@
+		) {
+			UserAccountsView()
+		}
+	// return PreferencePane
```
Should be removed.

sindresorhus commented on Example/UserAccountsView.swift:
```diff
@@ -0,0 +1,92 @@
+}
+
+/**
+The main view of "User Accounts" preference pane.
```
```suggestion
The main view of “User Accounts” preference pane.
```

sindresorhus commented:
Let's rename `PaneView` to `Pane` as no other view has `View`-suffix.

fredyshox commented:
@sindresorhus One issue with `init<T: View>(paneViews: [PaneView<T>])` in `PreferencesWindowController` came out, that as we store generic views in the array, all of them would need to have same generic type `T`.

I solved it by defining `PreferencePaneConvertible`, which acts as type eraser.

sindresorhus commented:
Can you fix the merge conflict?

sindresorhus commented:
Nice work, @fredyshox  🙌

hisaac commented:
I'm seeing some weird tab animation when using SwiftUI views. Any idea why that might be? #59
